src/main/java/www/purple/mixxy/controllers/LoginLogoutController.java

```diff
@@ -82,13 +58,69 @@ public Result logout(Context context) {
 
     }
 
-    public Result signup() {
-      return Results.TODO();
+	///////////////////////////////////////////////////////////////////////////
+	// Login
+	///////////////////////////////////////////////////////////////////////////
+    public Result login() {
+    	GoogleAuthHelper helper = new GoogleAuthHelper();
+    	return Results.redirect(helper.buildLoginUrl());
     }
     
-    public Result signupPost() {
-      return Results.TODO();
+    public Result validate(
+    		@Param("state") String state,
+    		@Param("code") String code,
+    		Context context) {
+    	
+    	GoogleAuthHelper helper = new GoogleAuthHelper();
+    	String data = "empty";
+    	String provider = state.substring(0, state.indexOf(';'));
+    	
+    	try {
+			data = helper.getUserInfoJson(code);
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+    	
+    	ObjectMapper mapper = new ObjectMapper();
+		try {
+
+			// Convert JSON string to Object
+			GoogleAuthResponse userdata = mapper.readValue(data, GoogleAuthResponse.class);
+			
+			// At this point we can login or signup user
+			String email = userdata.getEmail();
+			String username = email.substring(0, email.indexOf('@'));
```

Comment:
I envisioned the user picking a username after they register with a provider (i.e. username's not known until after they give OAuth consent).  Maybe this can be a sensible default?